sway-lib-std/src/ufp64.sw

```diff
@@ -0,0 +1,247 @@
+        self.value > other.value
+    }
+
+    pub fn lt(self, other: Self) -> bool {
+        self.value < other.value
+    }
+}
+
+impl core::ops::Add for UFP64 {
+    /// Add a UFP64 to a UFP64. Panics on overflow.
+    pub fn add(self, other: Self) -> Self {
+        UFP64 {
+            value: self.value + other.value
+        }
+    }
+}
+
+impl core::ops::Subtract for UFP64 {
+    /// Subtract a UFP64 from a UFP64. Panics of overflow.
+    pub fn subtract(self, other: Self) -> Self {
+        // If trying to subtract a larger number, panic.
+        assert(!(self.value < other.value));
+
+        UFP64 {
+            value: self.value - other.value
+        }
+    }
+}
+
+impl core::ops::Multiply for UFP64 {
+    /// Multiply a UFP64 with a UFP64. Panics of overflow.
+    pub fn multiply(self, other: Self) -> Self {
+
+        let s_value = ~U128::from(0, self.value);
+        let o_value = ~U128::from(0, other.value);
+
+        let s_mul_o = s_value * o_value;
+        let res_u128 = s_mul_o >> 32;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+
+        UFP64 {
+            value: res_u128.lower
+        }
+    }
+}
+
+impl core::ops::Divide for UFP64 {
+    /// Divide a UFP64 by a UFP64. Panics if divisor is zero.
+    pub fn divide(self, divisor: Self) -> Self {
+    let zero = ~UFP64::zero();
+        assert(divisor != zero);
+
+        let denominator = ~U128::from(0, ~Self::denominator());
+        let self_u128 = ~U128::from(0, self.value);
+        let divisor_u128 = ~U128::from(0, divisor.value);
+
+        let res_u128 = self_u128 * denominator / divisor_u128;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        UFP64 {
+            value: res_u128.lower
+        }
+    }
+}
+
+
+
+impl UFP64 {
+
+    pub fn from_uint(uint: u64) -> UFP64 {
+        UFP64 {
+            value: ~Self::denominator() * uint,
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn recip(number: UFP64) -> Self {
+        let one = ~UFP64::from_uint(1);
+        
+        let res = one / number;
+        res
+    }
+}
+
+impl UFP64 {
+
+    pub fn trunc(self) -> Self {
+        UFP64 {
+            value: (self.value >> 32) << 32
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn floor(self) -> Self {
+        return self.trunc();
+    }
+
+    pub fn fract(self) -> Self {
+        UFP64 {
+            value: (self.value << 32) >> 32
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn ceil(self) -> Self {
+        if self.fract().value != 0 {
+            let res = self.trunc() + ~UFP64::from_uint(1);
+            return res;
+        }
+        return self;
+    }
+}
+
+impl UFP64 {
+
+    pub fn round(self) -> Self {
+        let floor = self.floor();
+        let ceil = self.ceil();
+        let diff_self_floor = self - floor;
+        let diff_ceil_self = ceil - self;
+        if diff_self_floor < diff_ceil_self {
```

Comment:
```suggestion

        if diff_self_floor < diff_ceil_self {
```

Just a little easier to read. I'd also consider changing this into a ternary return but this is fine.

Reply:
Done